# **Project Name**    -



##### **Project Type**    - EDA/Regression/Classification/Unsupervised
##### **Contribution**    - Individual/Team
##### **Team Member 1 -**
##### **Team Member 2 -**
##### **Team Member 3 -**
##### **Team Member 4 -**

# **Project Summary -**

Write the summary here within 500-600 words.

# **GitHub Link -**

Provide your GitHub Link here.

# **Problem Statement**


**Write Problem Statement Here.**

# **General Guidelines** : -  

1.   Well-structured, formatted, and commented code is required.
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits.
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 15 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule.

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





6. You may add more ml algorithms for model creation. Make sure for each and every algorithm, the following format should be answered.


*   Explain the ML Model used and it's performance using Evaluation metric Score Chart.


*   Cross- Validation & Hyperparameter Tuning

*   Have you seen any improvement? Note down the improvement with updates Evaluation metric Score Chart.

*   Explain each evaluation metric's indication towards business and the business impact pf the ML model used.




















# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

print("Libraries imported successfully.")

Libraries imported successfully.


### Dataset Loading

In [2]:
# Load Dataset
# GitHub Raw URLs
users_url = "https://raw.githubusercontent.com/Sagarjain93/Machine-Learning-Portfolio/refs/heads/main/01_End_to_End/Travel_MLops_Capstone/data/raw_data/users.csv"
flights_url = "https://raw.githubusercontent.com/Sagarjain93/Machine-Learning-Portfolio/refs/heads/main/01_End_to_End/Travel_MLops_Capstone/data/raw_data/flights.csv"
hotels_url = "https://raw.githubusercontent.com/Sagarjain93/Machine-Learning-Portfolio/refs/heads/main/01_End_to_End/Travel_MLops_Capstone/data/raw_data/hotels.csv"

# Load datasets
users = pd.read_csv(users_url)
flights = pd.read_csv(flights_url)
hotels = pd.read_csv(hotels_url)

print("All three datasets loaded successfully.")

All three datasets loaded successfully.


### Dataset First View

In [3]:
# Dataset First Look
# Display first 5 records of Users dataset

users.head()

,code,company,name,gender,age
0,0,4You,Roy Braun,male,21
1,1,4You,Joseph Holsten,male,37
2,2,4You,Wilma Mcinnis,female,48
3,3,4You,Paula Daniel,female,23
4,4,4You,Patricia Carson,female,44


In [4]:
# Display last 5 records of Users dataset

users.tail()

,code,company,name,gender,age
1335,1335,Umbrella LTDA,Albert Garroutte,male,23
1336,1336,Umbrella LTDA,Kim Shores,female,40
1337,1337,Umbrella LTDA,James Gimenez,male,28
1338,1338,Umbrella LTDA,Viola Agosta,female,52
1339,1339,Umbrella LTDA,Paul Rodriguez,male,35


In [5]:
# Display first 5 records of Flights dataset

flights.head()

,travelCode,userCode,from,to,flightType,price,time,distance,agency,date
0,0,0,Recife (PE),Florianopolis (SC),firstClass,1434.38,1.76,676.53,FlyingDrops,09/26/2019
1,0,0,Florianopolis (SC),Recife (PE),firstClass,1292.29,1.76,676.53,FlyingDrops,09/30/2019
2,1,0,Brasilia (DF),Florianopolis (SC),firstClass,1487.52,1.66,637.56,CloudFy,10/03/2019
3,1,0,Florianopolis (SC),Brasilia (DF),firstClass,1127.36,1.66,637.56,CloudFy,10/04/2019
4,2,0,Aracaju (SE),Salvador (BH),firstClass,1684.05,2.16,830.86,CloudFy,10/10/2019


In [6]:
# Display last 5 records of Flights dataset

flights.tail()

,travelCode,userCode,from,to,flightType,price,time,distance,agency,date
271883,135941,1339,Campo Grande (MS),Florianopolis (SC),firstClass,1446.34,1.49,573.81,CloudFy,07/12/2020
271884,135942,1339,Florianopolis (SC),Natal (RN),economic,726.95,1.84,709.37,CloudFy,07/16/2020
271885,135942,1339,Natal (RN),Florianopolis (SC),economic,873.07,1.84,709.37,CloudFy,07/20/2020
271886,135943,1339,Florianopolis (SC),Rio de Janeiro (RJ),economic,313.62,1.21,466.30,CloudFy,07/23/2020
271887,135943,1339,Rio de Janeiro (RJ),Florianopolis (SC),economic,533.69,1.21,466.30,CloudFy,07/26/2020


In [7]:
# Display first 5 records of Hotels dataset

hotels.head()

,travelCode,userCode,name,place,days,price,total,date
0,0,0,Hotel A,Florianopolis (SC),4,313.02,1252.08,09/26/2019
1,2,0,Hotel K,Salvador (BH),2,263.41,526.82,10/10/2019
2,7,0,Hotel K,Salvador (BH),3,263.41,790.23,11/14/2019
3,11,0,Hotel K,Salvador (BH),4,263.41,1053.64,12/12/2019
4,13,0,Hotel A,Florianopolis (SC),1,313.02,313.02,12/26/2019


In [8]:
# Display last 5 records of Hotels dataset

hotels.tail()

,travelCode,userCode,name,place,days,price,total,date
40547,135938,1339,Hotel BP,Brasilia (DF),3,247.62,742.86,06/18/2020
40548,135939,1339,Hotel BP,Brasilia (DF),1,247.62,247.62,06/25/2020
40549,135940,1339,Hotel BW,Campo Grande (MS),3,60.39,181.17,07/02/2020
40550,135941,1339,Hotel BW,Campo Grande (MS),3,60.39,181.17,07/09/2020
40551,135942,1339,Hotel BD,Natal (RN),4,242.88,971.52,07/16/2020


In [9]:
# Display 5 random records from each dataset

print("Users - Random Sample")
display(users.sample(5))

print("\nFlights - Random Sample")
display(flights.sample(5))

print("\nHotels - Random Sample")
display(hotels.sample(5))

Users - Random Sample


,code,company,name,gender,age
317,317,4You,Adam Groome,male,54
581,581,Monsters CYA,Jennie Klein,female,35
837,837,Wonka Company,Harold Uren,male,50
659,659,Wonka Company,Juan Kloke,male,57
202,202,4You,Christopher Schroeder,male,34



Flights - Random Sample


,travelCode,userCode,from,to,flightType,price,time,distance,agency,date
217467,108733,1078,Natal (RN),Florianopolis (SC),firstClass,1554.92,1.84,709.37,Rainbow,06/03/2022
157300,78650,763,Natal (RN),Florianopolis (SC),premium,1212.58,1.84,709.37,Rainbow,06/23/2022
26497,13248,129,Salvador (BH),Aracaju (SE),firstClass,1520.02,2.16,830.86,Rainbow,10/23/2022
176905,88452,873,Florianopolis (SC),Natal (RN),premium,1128.51,1.84,709.37,CloudFy,12/24/2021
186037,93018,918,Recife (PE),Florianopolis (SC),economic,761.20,1.76,676.53,CloudFy,04/17/2020



Hotels - Random Sample


,travelCode,userCode,name,place,days,price,total,date
25227,84748,832,Hotel AF,Sao Paulo (SP),2,139.10,278.20,11/05/2020
248,894,8,Hotel BP,Brasilia (DF),3,247.62,742.86,02/11/2021
20877,69921,677,Hotel BP,Brasilia (DF),1,247.62,247.62,03/04/2021
2245,7537,72,Hotel BP,Brasilia (DF),3,247.62,742.86,10/22/2020
3256,10878,109,Hotel BD,Natal (RN),4,242.88,971.52,05/07/2020


### Dataset Rows & Columns count

In [10]:
# Dataset Rows & Columns count
print("Users:", users.shape)
print("Flights:", flights.shape)
print("Hotels:", hotels.shape)

Users: (1340, 5)
Flights: (271888, 10)
Hotels: (40552, 8)


### Dataset Information

In [11]:
# Dataset Info
# Display concise summary information for all three datasets

print("========== USERS DATASET ==========")
users.info()

print("\n========== FLIGHTS DATASET ==========")
flights.info()

print("\n========== HOTELS DATASET ==========")
hotels.info()

========== USERS DATASET ==========
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1340 entries, 0 to 1339
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   code     1340 non-null   int64 
 1   company  1340 non-null   object
 2   name     1340 non-null   object
 3   gender   1340 non-null   object
 4   age      1340 non-null   int64 
dtypes: int64(2), object(3)
memory usage: 52.5+ KB

========== FLIGHTS DATASET ==========
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271888 entries, 0 to 271887
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   travelCode  271888 non-null  int64  
 1   userCode    271888 non-null  int64  
 2   from        271888 non-null  object 
 3   to          271888 non-null  object 
 4   flightType  271888 non-null  object 
 5   price       271888 non-null  float64
 6   time        271888 non-null  float64
 7   distance    271

In [12]:
# Display data types of all columns

print("Users Data Types:")
display(users.dtypes)

print("\nFlights Data Types:")
display(flights.dtypes)

print("\nHotels Data Types:")
display(hotels.dtypes)

Users Data Types:


,0
code,int64
company,object
name,object
gender,object
age,int64



Flights Data Types:


,0
travelCode,int64
userCode,int64
from,object
to,object
flightType,object
price,float64
time,float64
distance,float64
agency,object
date,object



Hotels Data Types:


,0
travelCode,int64
userCode,int64
name,object
place,object
days,int64
price,float64
total,float64
date,object


#### Duplicate Values

In [13]:
# Dataset Duplicate Value Count
# Check duplicate records in each dataset

print("Duplicate rows in Users:", users.duplicated().sum())
print("Duplicate rows in Flights:", flights.duplicated().sum())
print("Duplicate rows in Hotels:", hotels.duplicated().sum())

Duplicate rows in Users: 0
Duplicate rows in Flights: 0
Duplicate rows in Hotels: 0


#### Missing Values/Null Values

In [14]:
# Missing Values/Null Values Count
# Check missing values in each dataset

print("========== USERS MISSING VALUES ==========")
display(users.isnull().sum())

print("\n========== FLIGHTS MISSING VALUES ==========")
display(flights.isnull().sum())

print("\n========== HOTELS MISSING VALUES ==========")
display(hotels.isnull().sum())

========== USERS MISSING VALUES ==========


,0
code,0
company,0
name,0
gender,0
age,0



========== FLIGHTS MISSING VALUES ==========


,0
travelCode,0
userCode,0
from,0
to,0
flightType,0
price,0
time,0
distance,0
agency,0
date,0



========== HOTELS MISSING VALUES ==========


,0
travelCode,0
userCode,0
name,0
place,0
days,0
price,0
total,0
date,0


In [15]:
# Visualizing the missing values

### What did you know about your dataset?

The dataset consists of three interconnected datasets related to travel and tourism: Users, Flights, and Hotels.

The Users dataset contains 1,340 records and 5 variables, providing information about users such as their company, name, gender, and age.

The Flights dataset contains 271,888 records and 10 variables. It provides information about flight transactions, including travel code, user code, source and destination, flight type, price, travel time, distance, agency, and date.

The Hotels dataset contains 40,552 records and 8 variables. It includes information about hotel bookings such as travel code, user code, hotel name, location, number of days, price per day, total price, and date.

The datasets are connected through identifiers such as userCode and travelCode, which allow us to relate users with their flight and hotel travel activities.

During the initial data-quality assessment, no missing values and no duplicate rows were found in any of the three datasets. The numerical and categorical variables are currently stored using appropriate basic pandas data types, although some variables such as date are stored as object/string values and will require further investigation and conversion during data preprocessing.

The datasets provide a useful foundation for developing three different machine learning components: flight price prediction, gender classification, and hotel recommendation.

## ***2. Understanding Your Variables***

In [16]:
# Dataset Columns
# Display columns of all three datasets

print("========== USERS COLUMNS ==========")
print(users.columns.tolist())

print("\n========== FLIGHTS COLUMNS ==========")
print(flights.columns.tolist())

print("\n========== HOTELS COLUMNS ==========")
print(hotels.columns.tolist())

========== USERS COLUMNS ==========
['code', 'company', 'name', 'gender', 'age']

========== FLIGHTS COLUMNS ==========
['travelCode', 'userCode', 'from', 'to', 'flightType', 'price', 'time', 'distance', 'agency', 'date']

========== HOTELS COLUMNS ==========
['travelCode', 'userCode', 'name', 'place', 'days', 'price', 'total', 'date']


In [17]:
# Dataset Describe
# Statistical summary of numerical variables

print("========== USERS DATASET ==========")
display(users.describe())

print("\n========== FLIGHTS DATASET ==========")
display(flights.describe())

print("\n========== HOTELS DATASET ==========")
display(hotels.describe())

========== USERS DATASET ==========


,code,age
count,1340.000000,1340.000000
mean,669.500000,42.742537
std,386.968991,12.869779
min,0.000000,21.000000
25%,334.750000,32.000000
50%,669.500000,42.000000
75%,1004.250000,54.000000
max,1339.000000,65.000000



========== FLIGHTS DATASET ==========


,travelCode,userCode,price,time,distance
count,271888.000000,271888.000000,271888.00000,271888.000000,271888.000000
mean,67971.500000,667.505495,957.37503,1.421147,546.955535
std,39243.724665,389.523127,362.31189,0.542541,208.851288
min,0.000000,0.000000,301.51000,0.440000,168.220000
25%,33985.750000,326.000000,672.66000,1.040000,401.660000
50%,67971.500000,659.000000,904.00000,1.460000,562.140000
75%,101957.250000,1011.000000,1222.24000,1.760000,676.530000
max,135943.000000,1339.000000,1754.17000,2.440000,937.770000



========== HOTELS DATASET ==========


,travelCode,userCode,days,price,total
count,40552.000000,40552.000000,40552.000000,40552.000000,40552.000000
mean,67911.794461,666.963726,2.499679,214.439554,536.229513
std,39408.199333,391.136794,1.119326,76.742305,319.331482
min,0.000000,0.000000,1.000000,60.390000,60.390000
25%,33696.750000,323.000000,1.000000,165.990000,247.620000
50%,67831.000000,658.000000,2.000000,242.880000,495.240000
75%,102211.250000,1013.000000,4.000000,263.410000,742.860000
max,135942.000000,1339.000000,4.000000,313.020000,1252.080000


### Variables Description

| Dataset | Variable     | Description                                    |
| ------- | ------------ | ---------------------------------------------- |
| Users   | `code`       | Unique identifier for the user                 |
| Users   | `company`    | Company associated with the user               |
| Users   | `name`       | Name of the user                               |
| Users   | `gender`     | Gender of the user                             |
| Users   | `age`        | Age of the user                                |
| Flights | `travelCode` | Identifier of the travel transaction           |
| Flights | `userCode`   | Identifier linking the flight to a user        |
| Flights | `from`       | Flight origin                                  |
| Flights | `to`         | Flight destination                             |
| Flights | `flightType` | Type/class of flight                           |
| Flights | `price`      | Flight price                                   |
| Flights | `time`       | Flight duration                                |
| Flights | `distance`   | Flight distance                                |
| Flights | `agency`     | Agency associated with the flight              |
| Flights | `date`       | Flight date                                    |
| Hotels  | `travelCode` | Identifier of the travel transaction           |
| Hotels  | `userCode`   | Identifier linking the hotel booking to a user |
| Hotels  | `name`       | Hotel name                                     |
| Hotels  | `place`      | Hotel location                                 |
| Hotels  | `days`       | Number of days of stay                         |
| Hotels  | `price`      | Price per day                                  |
| Hotels  | `total`      | Total hotel booking cost                       |
| Hotels  | `date`       | Hotel booking date                             |


### Check Unique Values for each variable.

In [18]:
# Check Unique Values for each variable.
# Unique values in Users dataset

print("Company values:")
print(users['company'].unique())

print("\nGender values:")
print(users['gender'].unique())

print("\nNumber of unique users:")
print(users['code'].nunique())

print("\nAge range:")
print("Minimum age:", users['age'].min())
print("Maximum age:", users['age'].max())

Company values:
['4You' 'Monsters CYA' 'Wonka Company' 'Acme Factory' 'Umbrella LTDA']

Gender values:
['male' 'female' 'none']

Number of unique users:
1340

Age range:
Minimum age: 21
Maximum age: 65


**1. Gender distribution**

In [19]:
# Check frequency of each gender category

print("Gender distribution:")
display(users['gender'].value_counts(dropna=False))

Gender distribution:


,count
gender,
male,452
female,448
none,440


2. Flight type

In [20]:
# Check flight type categories

print("Flight Type distribution:")
display(flights['flightType'].value_counts(dropna=False))

Flight Type distribution:


,count
flightType,
firstClass,116418
premium,78004
economic,77466


3. Agency

In [21]:
# Check agency categories

print("Agency distribution:")
display(flights['agency'].value_counts(dropna=False))

Agency distribution:


,count
agency,
Rainbow,116752
CloudFy,116378
FlyingDrops,38758


4. Company

In [22]:
# Check company categories

print("Company distribution:")
display(users['company'].value_counts(dropna=False))

Company distribution:


,count
company,
4You,453
Acme Factory,261
Wonka Company,237
Monsters CYA,195
Umbrella LTDA,194


5. Hotel places

In [23]:
# Check hotel place categories

print("Hotel place distribution:")
display(hotels['place'].value_counts(dropna=False))

Hotel place distribution:


,count
place,
Salvador (BH),5094
Rio de Janeiro (RJ),5029
Natal (RN),4829
Sao Paulo (SP),4828
Recife (PE),4467
Brasilia (DF),4437
Campo Grande (MS),4333
Aracaju (SE),4205
Florianopolis (SC),3330


6. Hotel names

In [24]:
# Check hotel name distribution

print("Hotel name distribution:")
display(hotels['name'].value_counts(dropna=False))

Hotel name distribution:


,count
name,
Hotel K,5094
Hotel CB,5029
Hotel BD,4829
Hotel AF,4828
Hotel AU,4467
Hotel BP,4437
Hotel BW,4333
Hotel Z,4205
Hotel A,3330


In [75]:
# Check whether every flight userCode exists in the Users dataset

missing_flight_users = ~flights['userCode'].isin(users['code'])

print("Flight records with unknown userCode:",
      missing_flight_users.sum())

Flight records with unknown userCode: 0


In [76]:
# Check whether every hotel userCode exists in the Users dataset

missing_hotel_users = ~hotels['userCode'].isin(users['code'])

print("Hotel records with unknown userCode:",
      missing_hotel_users.sum())

Hotel records with unknown userCode: 0


In [77]:
# Check travelCode overlap between flights and hotels

flight_travel_codes = set(flights['travelCode'])
hotel_travel_codes = set(hotels['travelCode'])

print("Flight travel codes:", len(flight_travel_codes))
print("Hotel travel codes:", len(hotel_travel_codes))
print("Common travel codes:", len(flight_travel_codes & hotel_travel_codes))

Flight travel codes: 135944
Hotel travel codes: 40552
Common travel codes: 40552


## 3. ***Data Wrangling***

**Validate userCode**

In [78]:
# Validate relationships between Users and Flights/Hotels

# Check whether every flight userCode exists in Users.code
invalid_flight_users = ~flights['userCode'].isin(users['code'])

# Check whether every hotel userCode exists in Users.code
invalid_hotel_users = ~hotels['userCode'].isin(users['code'])

print("Flight records with invalid userCode:", invalid_flight_users.sum())
print("Hotel records with invalid userCode:", invalid_hotel_users.sum())

Flight records with invalid userCode: 0
Hotel records with invalid userCode: 0


**Validate travelCode**

In [79]:
# Validate travelCode relationship between Flights and Hotels

flight_travel_codes = set(flights['travelCode'])
hotel_travel_codes = set(hotels['travelCode'])

common_travel_codes = flight_travel_codes.intersection(hotel_travel_codes)

print("Unique travelCodes in Flights:", len(flight_travel_codes))
print("Unique travelCodes in Hotels:", len(hotel_travel_codes))
print("Common travelCodes:", len(common_travel_codes))

print("\nHotel travelCodes not present in Flights:",
      len(hotel_travel_codes - flight_travel_codes))

Unique travelCodes in Flights: 135944
Unique travelCodes in Hotels: 40552
Common travelCodes: 40552

Hotel travelCodes not present in Flights: 0


**Understand the travelCode structure**

**Flights**

In [81]:
# Check number of flight records associated with each travelCode

flight_travel_counts = flights['travelCode'].value_counts()

print("Maximum flight records for one travelCode:",
      flight_travel_counts.max())

print("Number of travelCodes with more than one flight record:",
      (flight_travel_counts > 1).sum())

print("\nDistribution of number of flight records per travelCode:")
display(flight_travel_counts.value_counts().sort_index())

Maximum flight records for one travelCode: 2
Number of travelCodes with more than one flight record: 135944

Distribution of number of flight records per travelCode:


,count
count,
2,135944


**Hotels**

In [82]:
# Check number of hotel records associated with each travelCode

hotel_travel_counts = hotels['travelCode'].value_counts()

print("Maximum hotel records for one travelCode:",
      hotel_travel_counts.max())

print("Number of travelCodes with more than one hotel record:",
      (hotel_travel_counts > 1).sum())

print("\nDistribution of number of hotel records per travelCode:")
display(hotel_travel_counts.value_counts().sort_index())

Maximum hotel records for one travelCode: 1
Number of travelCodes with more than one hotel record: 0

Distribution of number of hotel records per travelCode:


,count
count,
1,40552


**Check hotel total**

In [83]:
# Validate hotel total price

calculated_total = hotels['price'] * hotels['days']

total_difference = (
    hotels['total'] - calculated_total
).round(2)

print("Records where total != price × days:",
      (total_difference.abs() > 0.01).sum())

print("Maximum absolute difference:",
      total_difference.abs().max())

Records where total != price × days: 0
Maximum absolute difference: 0.0


**Check invalid numerical values**

In [84]:
# Check for invalid or non-positive numerical values

print("========== USERS ==========")
print("Age <= 0:", (users['age'] <= 0).sum())

print("\n========== FLIGHTS ==========")
print("Price <= 0:", (flights['price'] <= 0).sum())
print("Time <= 0:", (flights['time'] <= 0).sum())
print("Distance <= 0:", (flights['distance'] <= 0).sum())

print("\n========== HOTELS ==========")
print("Days <= 0:", (hotels['days'] <= 0).sum())
print("Price <= 0:", (hotels['price'] <= 0).sum())
print("Total <= 0:", (hotels['total'] <= 0).sum())

========== USERS ==========
Age <= 0: 0

========== FLIGHTS ==========
Price <= 0: 0
Time <= 0: 0
Distance <= 0: 0

========== HOTELS ==========
Days <= 0: 0
Price <= 0: 0
Total <= 0: 0


**Validate dates**

In [85]:
# Validate date columns without modifying the original datasets

flight_dates = pd.to_datetime(
    flights['date'],
    format='%m/%d/%Y',
    errors='coerce'
)

hotel_dates = pd.to_datetime(
    hotels['date'],
    format='%m/%d/%Y',
    errors='coerce'
)

print("Invalid flight dates:", flight_dates.isna().sum())
print("Invalid hotel dates:", hotel_dates.isna().sum())

print("\nFlight date range:")
print(flight_dates.min(), "to", flight_dates.max())

print("\nHotel date range:")
print(hotel_dates.min(), "to", hotel_dates.max())

Invalid flight dates: 0
Invalid hotel dates: 0

Flight date range:
2019-09-26 00:00:00 to 2023-07-24 00:00:00

Hotel date range:
2019-09-26 00:00:00 to 2023-07-13 00:00:00


**Investigate gender = none**

In [86]:
# Inspect users whose gender is recorded as 'none'

none_gender_users = users[users['gender'] == 'none']

print("Number of users with gender = 'none':",
      len(none_gender_users))

display(none_gender_users.head(10))

Number of users with gender = 'none': 440


,code,company,name,gender,age
5,5,4You,Trina Thomas,none,47
8,8,4You,Jack Sabo,none,41
9,9,4You,Debbie Helms,none,35
13,13,4You,Irene Tucker,none,36
17,17,4You,Victor Tribbett,none,25
18,18,4You,Tanya Orourke,none,65
31,31,4You,Christopher Elliott,none,40
35,35,4You,Helen Selby,none,34
36,36,4You,Anna Rodriguez,none,56
37,37,4You,Denise Newton,none,27


In [87]:
# Check whether users with gender = 'none' have flight records

none_user_codes = set(none_gender_users['code'])

none_gender_flights = flights[
    flights['userCode'].isin(none_user_codes)
]

print("Flight records associated with gender = 'none' users:",
      len(none_gender_flights))

print("Unique users with flight records:",
      none_gender_flights['userCode'].nunique())

Flight records associated with gender = 'none' users: 89060
Unique users with flight records: 437


In [88]:
# Check whether users with gender = 'none' have hotel records

none_gender_hotels = hotels[
    hotels['userCode'].isin(none_user_codes)
]

print("Hotel records associated with gender = 'none' users:",
      len(none_gender_hotels))

print("Unique users with hotel records:",
      none_gender_hotels['userCode'].nunique())

Hotel records associated with gender = 'none' users: 13277
Unique users with hotel records: 425


**Convert dates**

In [89]:
# Convert date columns to datetime

flights['date'] = pd.to_datetime(
    flights['date'],
    format='%m/%d/%Y'
)

hotels['date'] = pd.to_datetime(
    hotels['date'],
    format='%m/%d/%Y'
)

print("Flights date dtype:", flights['date'].dtype)
print("Hotels date dtype:", hotels['date'].dtype)

Flights date dtype: datetime64[ns]
Hotels date dtype: datetime64[ns]


In [90]:
print("Flights date range:")
print(flights['date'].min(), "to", flights['date'].max())

print("\nHotels date range:")
print(hotels['date'].min(), "to", hotels['date'].max())

Flights date range:
2019-09-26 00:00:00 to 2023-07-24 00:00:00

Hotels date range:
2019-09-26 00:00:00 to 2023-07-13 00:00:00


**Flights associated with none-gender users**

In [91]:
# Check whether users with gender = 'none' have flight records

none_user_codes = set(none_gender_users['code'])

none_gender_flights = flights[
    flights['userCode'].isin(none_user_codes)
]

print("Flight records associated with gender = 'none' users:",
      len(none_gender_flights))

print("Unique users with flight records:",
      none_gender_flights['userCode'].nunique())

Flight records associated with gender = 'none' users: 89060
Unique users with flight records: 437


**Hotels associated with none-gender users**

In [92]:
# Check whether users with gender = 'none' have hotel records

none_gender_hotels = hotels[
    hotels['userCode'].isin(none_user_codes)
]

print("Hotel records associated with gender = 'none' users:",
      len(none_gender_hotels))

print("Unique users with hotel records:",
      none_gender_hotels['userCode'].nunique())

Hotel records associated with gender = 'none' users: 13277
Unique users with hotel records: 425


**Check categorical cleanliness**

In [93]:
# Check for leading/trailing whitespace in categorical columns

categorical_columns = {
    'users': ['company', 'name', 'gender'],
    'flights': ['from', 'to', 'flightType', 'agency'],
    'hotels': ['name', 'place']
}

for dataset_name, columns in categorical_columns.items():
    df = {
        'users': users,
        'flights': flights,
        'hotels': hotels
    }[dataset_name]

    print(f"\n========== {dataset_name.upper()} ==========")

    for col in columns:
        whitespace_count = (
            df[col].astype(str).str.strip() != df[col].astype(str)
        ).sum()

        print(f"{col}: {whitespace_count} values with leading/trailing whitespace")


========== USERS ==========
company: 0 values with leading/trailing whitespace
name: 0 values with leading/trailing whitespace
gender: 0 values with leading/trailing whitespace

========== FLIGHTS ==========
from: 0 values with leading/trailing whitespace
to: 0 values with leading/trailing whitespace
flightType: 0 values with leading/trailing whitespace
agency: 0 values with leading/trailing whitespace

========== HOTELS ==========
name: 0 values with leading/trailing whitespace
place: 0 values with leading/trailing whitespace


**Check for outliers**

**Users — Age**

In [94]:
# Function to calculate IQR-based outliers

def outlier_summary(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = ((df[column] < lower_bound) |
                (df[column] > upper_bound)).sum()

    return Q1, Q3, IQR, lower_bound, upper_bound, outliers


Q1, Q3, IQR, lower, upper, outliers = outlier_summary(users, 'age')

print("Users - Age")
print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower)
print("Upper Bound:", upper)
print("Potential Outliers:", outliers)

Users - Age
Q1: 32.0
Q3: 54.0
IQR: 22.0
Lower Bound: -1.0
Upper Bound: 87.0
Potential Outliers: 0


**Flights**

In [95]:
# Check potential outliers in flight numerical variables

for col in ['price', 'time', 'distance']:
    Q1, Q3, IQR, lower, upper, outliers = outlier_summary(flights, col)

    print(f"\nFlights - {col}")
    print("Q1:", Q1)
    print("Q3:", Q3)
    print("IQR:", IQR)
    print("Lower Bound:", lower)
    print("Upper Bound:", upper)
    print("Potential Outliers:", outliers)


Flights - price
Q1: 672.66
Q3: 1222.24
IQR: 549.58
Lower Bound: -151.71000000000015
Upper Bound: 2046.6100000000001
Potential Outliers: 0

Flights - time
Q1: 1.04
Q3: 1.76
IQR: 0.72
Lower Bound: -0.040000000000000036
Upper Bound: 2.84
Potential Outliers: 0

Flights - distance
Q1: 401.66
Q3: 676.53
IQR: 274.86999999999995
Lower Bound: -10.644999999999925
Upper Bound: 1088.835
Potential Outliers: 0


**Hotels**

In [96]:
# Check potential outliers in hotel numerical variables

for col in ['days', 'price', 'total']:
    Q1, Q3, IQR, lower, upper, outliers = outlier_summary(hotels, col)

    print(f"\nHotels - {col}")
    print("Q1:", Q1)
    print("Q3:", Q3)
    print("IQR:", IQR)
    print("Lower Bound:", lower)
    print("Upper Bound:", upper)
    print("Potential Outliers:", outliers)


Hotels - days
Q1: 1.0
Q3: 4.0
IQR: 3.0
Lower Bound: -3.5
Upper Bound: 8.5
Potential Outliers: 0

Hotels - price
Q1: 165.99
Q3: 263.41
IQR: 97.42000000000002
Lower Bound: 19.859999999999985
Upper Bound: 409.5400000000001
Potential Outliers: 0

Hotels - total
Q1: 247.62
Q3: 742.86
IQR: 495.24
Lower Bound: -495.24
Upper Bound: 1485.72
Potential Outliers: 0


In [97]:
# Check number of flight and hotel records per user

flight_user_counts = flights['userCode'].value_counts()
hotel_user_counts = hotels['userCode'].value_counts()

print("========== FLIGHT RECORDS PER USER ==========")
print("Minimum:", flight_user_counts.min())
print("Maximum:", flight_user_counts.max())
print("Average:", round(flight_user_counts.mean(), 2))

print("\n========== HOTEL RECORDS PER USER ==========")
print("Minimum:", hotel_user_counts.min())
print("Maximum:", hotel_user_counts.max())
print("Average:", round(hotel_user_counts.mean(), 2))

========== FLIGHT RECORDS PER USER ==========
Minimum: 2
Maximum: 400
Average: 203.66

========== HOTEL RECORDS PER USER ==========
Minimum: 1
Maximum: 76
Average: 30.96


### What all manipulations have you done and insights you found?

### What all manipulations have you done and insights you found?

#### Data Manipulations / Data Cleaning Performed

1. Loaded and examined the Users, Flights, and Hotels datasets.

2. Checked the shape, data types, structure, and basic statistical properties of all three datasets.

3. Checked for missing values in all columns. No missing values were found in the three datasets.

4. Checked for duplicate rows. No duplicate rows were found in any of the datasets.

5. Examined unique values and frequency distributions of important categorical variables such as gender, company, flight type, agency, hotel name, and hotel location.

6. Validated the relationships between datasets using `userCode` and `travelCode`.
   - All flight and hotel `userCode` values correspond to valid users.
   - Every hotel `travelCode` is also present in the Flights dataset.
   - Each `travelCode` has exactly two flight records.
   - Each hotel `travelCode` has exactly one hotel record.

7. Investigated the meaning of `travelCode`. Since every flight `travelCode` appears twice, it was treated as a trip/transaction identifier rather than a unique row identifier. Therefore, the repeated flight records were not incorrectly removed as duplicates.

8. Checked the number of travel records associated with each user.
   - Each user has at least 2 flight records and at most 400 flight records.
   - Each user has at least 1 hotel record and at most 76 hotel records.
   - This confirms that `userCode` represents a customer relationship across multiple travel transactions.

9. Investigated the `gender = 'none'` category. These records were retained in the master Users dataset because they represent existing users with associated travel activity. They will be handled separately when preparing the gender classification model rather than assigning an artificial gender label.

10. Converted the `date` columns in the Flights and Hotels datasets from string/object format into proper datetime format to enable date-based analysis and feature engineering.

11. Validated the date columns after conversion. No invalid dates were found.

12. Checked categorical columns for leading or trailing whitespace. No unwanted whitespace was found.

13. Checked numerical variables for invalid/non-positive values such as negative prices, distances, travel times, ages, hotel days, and hotel totals. No invalid values were found.

14. Performed IQR-based outlier analysis on relevant numerical variables including age, flight price, time, distance, hotel days, hotel price, and hotel total. No potential outliers were detected using the IQR criterion, so no outlier removal or capping was performed.

15. Validated the Hotels dataset calculation by checking whether `total = price × days`. No mismatches were found.

#### Key Insights Found

- The datasets contain three interconnected components: users, flights, and hotels, linked primarily through `userCode` and `travelCode`.

- The Users dataset contains 1,340 customers, while the Flights and Hotels datasets contain multiple transactions for the same users.

- The Flights dataset contains 271,888 records, and every flight `travelCode` occurs exactly twice, indicating that a trip can contain two flight records.

- The Hotels dataset contains 40,552 records, with exactly one hotel record associated with each hotel `travelCode`.

- Users have different levels of travel activity. Flight records per user range from 2 to 400, while hotel records range from 1 to 76.

- User ages range from 21 to 65 years, with an average age of approximately 42.7 years.

- The Users dataset contains three gender categories: male, female, and none. The `none` category represents users whose gender is not specified.

- Flights contain three flight classes/types: economic, premium, and firstClass.

- Flights are operated through three agencies: Rainbow, CloudFy, and FlyingDrops.

- Hotel records cover nine different hotel names and corresponding locations.

- There are no missing values, duplicate rows, invalid dates, invalid numerical values, or IQR-based outliers requiring treatment.

- The consistency check between hotel `price`, `days`, and `total` showed that the hotel total is correctly calculated for all records.

- Overall, the datasets are structurally clean and consistent, so extensive data-cleaning operations were not required. The main preprocessing performed so far has been data validation, relationship validation, date conversion, and investigation of potentially important categorical values.

## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

#### Chart - 1

In [26]:
# Chart - 1 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 2

In [27]:
# Chart - 2 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 3

In [28]:
# Chart - 3 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 4

In [29]:
# Chart - 4 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 5

In [30]:
# Chart - 5 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 6

In [31]:
# Chart - 6 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 7

In [32]:
# Chart - 7 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 8

In [33]:
# Chart - 8 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 9

In [34]:
# Chart - 9 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 10

In [35]:
# Chart - 10 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 11

In [36]:
# Chart - 11 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 12

In [37]:
# Chart - 12 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 13

In [38]:
# Chart - 13 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 14 - Correlation Heatmap

In [39]:
# Correlation Heatmap visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

#### Chart - 15 - Pair Plot

In [40]:
# Pair Plot visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

## ***5. Hypothesis Testing***

### Based on your chart experiments, define three hypothetical statements from the dataset. In the next three questions, perform hypothesis testing to obtain final conclusion about the statements through your code and statistical testing.

Answer Here.

### Hypothetical Statement - 1

#### 1. State Your research hypothesis as a null hypothesis and alternate hypothesis.

Answer Here.

#### 2. Perform an appropriate statistical test.

In [41]:
# Perform Statistical Test to obtain P-Value

##### Which statistical test have you done to obtain P-Value?

Answer Here.

##### Why did you choose the specific statistical test?

Answer Here.

### Hypothetical Statement - 2

#### 1. State Your research hypothesis as a null hypothesis and alternate hypothesis.

Answer Here.

#### 2. Perform an appropriate statistical test.

In [42]:
# Perform Statistical Test to obtain P-Value

##### Which statistical test have you done to obtain P-Value?

Answer Here.

##### Why did you choose the specific statistical test?

Answer Here.

### Hypothetical Statement - 3

#### 1. State Your research hypothesis as a null hypothesis and alternate hypothesis.

Answer Here.

#### 2. Perform an appropriate statistical test.

In [43]:
# Perform Statistical Test to obtain P-Value

##### Which statistical test have you done to obtain P-Value?

Answer Here.

##### Why did you choose the specific statistical test?

Answer Here.

## ***6. Feature Engineering & Data Pre-processing***

### 1. Handling Missing Values

In [44]:
# Handling Missing Values & Missing Value Imputation

#### What all missing value imputation techniques have you used and why did you use those techniques?

Answer Here.

### 2. Handling Outliers

In [45]:
# Handling Outliers & Outlier treatments

##### What all outlier treatment techniques have you used and why did you use those techniques?

Answer Here.

### 3. Categorical Encoding

In [46]:
# Encode your categorical columns

#### What all categorical encoding techniques have you used & why did you use those techniques?

Answer Here.

### 4. Textual Data Preprocessing
(It's mandatory for textual dataset i.e., NLP, Sentiment Analysis, Text Clustering etc.)

#### 1. Expand Contraction

In [47]:
# Expand Contraction

#### 2. Lower Casing

In [48]:
# Lower Casing

#### 3. Removing Punctuations

In [49]:
# Remove Punctuations

#### 4. Removing URLs & Removing words and digits contain digits.

In [50]:
# Remove URLs & Remove words and digits contain digits

#### 5. Removing Stopwords & Removing White spaces

In [51]:
# Remove Stopwords

In [52]:
# Remove White spaces

#### 6. Rephrase Text

In [53]:
# Rephrase Text

#### 7. Tokenization

In [54]:
# Tokenization

#### 8. Text Normalization

In [55]:
# Normalizing Text (i.e., Stemming, Lemmatization etc.)

##### Which text normalization technique have you used and why?

Answer Here.

#### 9. Part of speech tagging

In [56]:
# POS Taging

#### 10. Text Vectorization

In [57]:
# Vectorizing Text

##### Which text vectorization technique have you used and why?

Answer Here.

### 4. Feature Manipulation & Selection

#### 1. Feature Manipulation

In [58]:
# Manipulate Features to minimize feature correlation and create new features

#### 2. Feature Selection

In [59]:
# Select your features wisely to avoid overfitting

##### What all feature selection methods have you used  and why?

Answer Here.

##### Which all features you found important and why?

Answer Here.

### 5. Data Transformation

#### Do you think that your data needs to be transformed? If yes, which transformation have you used. Explain Why?

In [60]:
# Transform Your data

### 6. Data Scaling

In [61]:
# Scaling your data

##### Which method have you used to scale you data and why?

### 7. Dimesionality Reduction

##### Do you think that dimensionality reduction is needed? Explain Why?

Answer Here.

In [62]:
# DImensionality Reduction (If needed)

##### Which dimensionality reduction technique have you used and why? (If dimensionality reduction done on dataset.)

Answer Here.

### 8. Data Splitting

In [63]:
# Split your data to train and test. Choose Splitting ratio wisely.

##### What data splitting ratio have you used and why?

Answer Here.

### 9. Handling Imbalanced Dataset

##### Do you think the dataset is imbalanced? Explain Why.

Answer Here.

In [64]:
# Handling Imbalanced Dataset (If needed)

##### What technique did you use to handle the imbalance dataset and why? (If needed to be balanced)

Answer Here.

## ***7. ML Model Implementation***

### ML Model - 1

In [65]:
# ML Model - 1 Implementation

# Fit the Algorithm

# Predict on the model

#### 1. Explain the ML Model used and it's performance using Evaluation metric Score Chart.

In [66]:
# Visualizing evaluation Metric Score chart

#### 2. Cross- Validation & Hyperparameter Tuning

In [67]:
# ML Model - 1 Implementation with hyperparameter optimization techniques (i.e., GridSearch CV, RandomSearch CV, Bayesian Optimization etc.)

# Fit the Algorithm

# Predict on the model

##### Which hyperparameter optimization technique have you used and why?

Answer Here.

##### Have you seen any improvement? Note down the improvement with updates Evaluation metric Score Chart.

Answer Here.

### ML Model - 2

#### 1. Explain the ML Model used and it's performance using Evaluation metric Score Chart.

In [68]:
# Visualizing evaluation Metric Score chart

#### 2. Cross- Validation & Hyperparameter Tuning

In [69]:
# ML Model - 1 Implementation with hyperparameter optimization techniques (i.e., GridSearch CV, RandomSearch CV, Bayesian Optimization etc.)

# Fit the Algorithm

# Predict on the model

##### Which hyperparameter optimization technique have you used and why?

Answer Here.

##### Have you seen any improvement? Note down the improvement with updates Evaluation metric Score Chart.

Answer Here.

#### 3. Explain each evaluation metric's indication towards business and the business impact pf the ML model used.

Answer Here.

### ML Model - 3

In [70]:
# ML Model - 3 Implementation

# Fit the Algorithm

# Predict on the model

#### 1. Explain the ML Model used and it's performance using Evaluation metric Score Chart.

In [71]:
# Visualizing evaluation Metric Score chart

#### 2. Cross- Validation & Hyperparameter Tuning

In [72]:
# ML Model - 3 Implementation with hyperparameter optimization techniques (i.e., GridSearch CV, RandomSearch CV, Bayesian Optimization etc.)

# Fit the Algorithm

# Predict on the model

##### Which hyperparameter optimization technique have you used and why?

Answer Here.

##### Have you seen any improvement? Note down the improvement with updates Evaluation metric Score Chart.

Answer Here.

### 1. Which Evaluation metrics did you consider for a positive business impact and why?

Answer Here.

### 2. Which ML model did you choose from the above created models as your final prediction model and why?

Answer Here.

### 3. Explain the model which you have used and the feature importance using any model explainability tool?

Answer Here.

## ***8.*** ***Future Work (Optional)***

### 1. Save the best performing ml model in a pickle file or joblib file format for deployment process.


In [73]:
# Save the File

### 2. Again Load the saved model file and try to predict unseen data for a sanity check.


In [74]:
# Load the File and predict unseen data.

### ***Congrats! Your model is successfully created and ready for deployment on a live server for a real user interaction !!!***

# **Conclusion**

Write the conclusion here.

### ***Hurrah! You have successfully completed your Machine Learning Capstone Project !!!***